internal/k8s/vcf_controller.go

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2020-2021 VMware, Inc.
+ * All Rights Reserved.
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*   http://www.apache.org/licenses/LICENSE-2.0
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package k8s
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/internal/lib"
+	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/utils"
+
+	"github.com/vmware/alb-sdk/go/clients"
+	"github.com/vmware/alb-sdk/go/session"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+)
+
+func (c *AviController) AddNCPBootstrapEventHandler(k8sinfo K8sinformers, stopCh <-chan struct{}, startSyncCh chan struct{}) {
+	cs := k8sinfo.Cs
+	utils.AviLog.Debugf("Creating event broadcaster for NCP Bootstrap CRD")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(utils.AviLog.Debugf)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: cs.CoreV1().Events("")})
+	NCPBootstrapHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			utils.AviLog.Infof("NCP Bootstrap ADD Event")
+			if c.ValidBSData() && startSyncCh != nil {
+				c.UpdateAviSecret()
+				startSyncCh <- struct{}{}
+				startSyncCh = nil
+			}
+		},
+		UpdateFunc: func(old, obj interface{}) {
+			utils.AviLog.Infof("NCP Bootstrap Update Event")
+			if c.ValidBSData() && startSyncCh != nil {
+				c.UpdateAviSecret()
+				startSyncCh <- struct{}{}
+				startSyncCh = nil
+			}
+		},
+	}
```

Comment:
Is there a reason you are missing the `DeleteFunc`? If the bootstrap CR is deleted, wouldn't you take some action?

Reply:
Currently we dont have any action If the bootstrap CR is deleted. If the avi-secret is valid, AKO would continue to work.